### ISSUE TITLE
Missing cell voltages

### ISSUE THREAD

**Issue by Kristo-R** [Psynosaur/esphome-deye-bms#3]
Hi,

I can see all values from PCS and from InterCan but no individual cell voltages.
Seems like there is something with "extended_id" because cell voltages are only missing(and Cell_high/Cell_low).
And with only InterCan i can see some Pack2 data but no Pack1 data(I can see it with PCS so thats not important)
I have SE-G5.1Pro-B not SE-G5.3

Any suggestions?

**Comment by HA1Andrzej** [comment]
@Kristo-R  Do you have a solution? I too have a problem with the voltage of individual cells.

**Comment by asmcc** [comment]
I can not see these described telegrams, too:

0x02098001
0x04008001
0x04018001
0x04028001
0x04038001

Tested with CAN-Monitor-3000.exe on DEYE RW-M6.1 Battery.

I assume, that is a battery / firmware or setting specific issue.

But let's wait for the answer from Psynosaur. He is the best DEYE expert after all his experiments.

**Comment by Psynosaur** [comment]
Hey there

Sorry for the delay, 

I started my reverse engineering of these batteries purely from using CAN-Monitor-3000.

If you're connected to the PCS there should be summary voltages for the array/battery.

This would show cell high and low of either the array(pack) or the single battery.

If there is no extended frames for cell voltage on InterCan bus then it either uses a different baud rate or the firmware is out of date?

I don't own these models, and DEYE might have changed their way of communication (which I doubt tbh).

Please try different baud rates and see if that unveiled additional frames?

Regards

**Comment by asmcc** [comment]
@Psynosaur: First, if you can see something on the CAN, then your bitrate is correct. I know, how CAN works, and  you should fancy me, it is fact. Because the coverage of CAN with control sums and other things is so, that either you see all or you see nothing. The question of me and of other users here was explicitly about these 5 CAN telegrams. But I was hard-working on this topic the whole time and have additional information about it
Second, as far, as I know all the here named batteries (5.1, 5.3, 6.1) have the same CAN protocolls on both CANs, on PCCAN and on InternCAN. Maybe some small differencies between the models are given, but I think, that 90-95% of CAN telegrams and their content are the same
Third, I solved the issue by my battery. But you need additional equipment for it. You need the so named "LowVoltageBmsUpperComputer" software from DEYE (versions 1.17 and 1.18 are given) and the RED ONE "Canalyst II" CAN adapter. You need this very expensive adapter, since the DEYE software works only with this adapter and not with cheaper UART-based adapter, which you can use with other software like CAN-Monitor-3000. This "CANALYST II" adapter comes from China and costs about 50 ... 70 EUR in Germany. You can order it also over Amazon, but delivery time for Europe is about 2-3 weeks. For 90 EUR you gain the adapter in 2-3 days with amazon prime in Europe. But it was too expensive for me. With this software and with this adapter you can parametrise the battery, flash new firmware and other things. "CANALYST II" has usually two CANs, but you can not use it at the same time with above described DEYE software, because of bug in this software. You can either connect InterCAN, see all cell voltages and edit most parameters and options or you can connect PCCAN and you will see the information about the whole battery. I assume, that the DEYE software sends a special CAN telegram on the InterCAN, which "wake up" the 5 CAN telegrams with cell voltages. Anyway, I could not see these 5 telegrams before I started the DEYE software and can see these 5 CAN telegrams thereafter. This special "wake up" CAN telegram is not documented in your Excel sheet with all the information. I have a sneaking suspicion about the extended ID of this CAN telegram, but I must check, if it really this telegram. If we know it exactly, then expensive "Canalyst II" adapter and DEYE software will not be needed to activate it. You can send this telegram using "CAN-Monitor-3000", too. Furthermore, I do not know WHY these CAN telegrams were not sended some time (going in "sleep mode"), but stays "activated" if you "wake up" it unique. And I do not know, how long there stays activated and what is the criterium for the "sleep mode".
Fourth, as you can see, I continued your work and the work of Adminius. I am writing actually a driver for my VICTRON inverter to communicate directly with DEYE battery over both CANs. Therefore your and Adminius pre-works were very important for me. I enlarged and continued it further and plan to publish my results on github as documentation and then my driver for VICTRON inverter as Python script. But both only, if it is ready. Since 2-3 days communication between the DEYE Battery and VICTRON Inverter works using PCCAN. But that is only PCCAN part of Adminius. If that works, I will program the InterCAN, to read the cell voltages.

Best regards,

asmcc

**Comment by asmcc** [comment]
As anounced above, I created my own documentation project on GITHUB for my own 48VDC project. It consists links to all projects, which I found in the internet. I created also a "colored" table for PCSCAN to best overview of all messages and alarm bits. Additionaly I converted your excel table to PDF, too. My first Python script is also posted on this documentation repository as preview. And it works now using both PCSCAN and INTERCAN.

https://github.com/asmcc/my48vdc

Best regards,

asmcc

**Comment by Psynosaur** [comment]
> Hi,
> 
> I can see all values from PCS and from InterCan but no individual cell voltages. Seems like there is something with "extended_id" because cell voltages are only missing(and Cell_high/Cell_low). And with only InterCan i can see some Pack2 data but no Pack1 data(I can see it with PCS so thats not important) I have SE-G5.1Pro-B not SE-G5.3
> 
> Any suggestions?

it seems there is a heartbeat frame we need to send.

I shall find it and implement it in InterCAN or PCS wherever it is found.

Sorry for delay, been rather busy

**Comment by asmcc** [comment]
I do not know, if this heartbeat really required. After I connected my DEYE battery to the DEYE tool only on-time, these messages will be sent automatically without a heartbeat. But as I run the DEYE software, I saw 1-2 messages, which are sent periodically from DEYE software to the battery.
I'll keep watching it and inform you here, if I will know more detail.

**Comment by Psynosaur** [comment]
Good day

Indeed it was a heartbeat frame to get it going from a non reporting state

Extended frame ID 0x18A2FF80 DLC=>1 Data=> 01 every 1 second

Extended frame ID 0x18A20080 DLC=>1 Data=> 01 only sent when connecting via Deye app

Cell voltages are back without the need to use the Deye app to activate.

![Image](https://github.com/user-attachments/assets/97201d03-2705-4042-9711-2cde7df4adac)

**Comment by Psynosaur** [comment]
Code has been updated 🚀

**Comment by Psynosaur** [comment]
Hi

I have now tested this multiple times, manually restarting the batteries by using the Switch on the front panel.

Once the batteries start up PCS information is available, but cell voltages are missing.

Then I restart the ESP32 and then information for InterCAN returns, after sending the `wake up` frame once.

This can be observed below via Grafana.

![Image](https://github.com/user-attachments/assets/c5fce233-4919-4eba-a7db-6556c36adde2)

This issue should now be resolved

**Comment by asmcc** [comment]
Yes, sending this can frame once is the solution! Therefore I was a little confused, as you said "Heartbeat" in the post before. Yes, the term "heartbeat" I know also from CanOPEN. But CanOPEN is the industrial "dialect" of CAN. That is a mix from other industrial bus protocolls and CAN. The "original" automitive CAN do not have "heartbeat", since all messages were sent cyclic. And in this case we have more from "original" automotive CAN as from industrial dialect.
But "we have more from" does not mean, that we are complettely dialectless. As you investigated and documented in your excel table, at least the parametrisation is so constructed, that we have here a classic interrogator-responder play. You send first a "question" and gain a "response", but once, not cyclic.
About cell voltages an this special extended CAN frames, which are sent by deye software. Thanks, that you observed it so exactly and found this difference between once one and cyclyc one frame. I assume it too, that this once frame serves as activation for further sending of cell voltages by the batteries. The frame, which is sending cyclic by DEYE software may mean "Hello, I am the DEYE software! Give me please more debug parameters, if I ask you" or something.
About the further "dialect" and about the "structure" of this extended CAN messages I fount following information in the internet:
First, I foung, that "command" and "respose" messages are so structured, that two last bytes are cross-changed by sender and response:
`command 0 x 18 94 01 40`
`response  0 x 18 94 40 01`
Second, I found in the internet a discription about this "dialect". Here an example for the "command" above:
`[Priority=18] [Command=94][BMS ID=01][UpLinkID=40]`
Now, we can "translate" your both messages above using this systematic:
`[Priority=18][Command=A2][BMS ID=FF][UplinkID=80]`
or
`[Priority=18][Command=A2][BMS ID=00][UplinkID=80]`
I assume, that 00 or FF are broadcast "addresses" or IDs, to which DEYE software will "speak". 80 is in my opinion the address for the "master" (inverter or DEYE software). "Master" means here not the classical master-slave alliance, but something similar.
We have here also a mix between all possible variants / dialects in the communication.

**Comment by Psynosaur** [comment]
> is it possible to get InterCAN messages with "Wakeup" messages without having a connection to PCS? The reason why i'm asking, i want to read BMS information via RS485 and InterCAN (Cell Voltage etc. via CAN), both on the same ESP32...

Hi

you can read the cell information from RS485 without the need for a CAN bus connection.

**Comment by Psynosaur** [comment]
You'll need to send `0xAA, 0x55, 0x01, 0x04, 0x00, 0x03, 0x70, 0x0D, 0x0A` to the serial bus though.

**Comment by asmcc** [comment]
Yes, this "WakeUp" is sent over InernCAN. But I do not know, if the ESP can run without the PCSCAN. Maybe you should comment out the whole PCSCAN from the code

**Comment by kaybin85** [comment]
> Hi
> 
> you can read the cell information from RS485 without the need for a CAN bus connection.

via RS485/Modbus it's not possible to fetch registers greater 1000 (e.g. pack1: 10054 min / 10055 max cell voltage); this is only possible when LSW3 is connected to the inverter via serial - but this Stick is very unreliable.

**Comment by Psynosaur** [comment]
> > Hi
> > you can read the cell information from RS485 without the need for a CAN bus connection.
> 
> via RS485/Modbus it's not possible to fetch registers greater 1000 (e.g. pack1: 10054 min / 10055 max cell voltage); this is only possible when LSW3 is connected to the inverter via serial - but this Stick is very unreliable.

Uhm, what model of battery do you have?

use a RJ45 splitter, so that CAN comms is maintained and you can sniff what you want.

here is a screenshot of the RS485 app from Deye, whilst CAN bus was chatting with the inverter over PCS.

![Image](https://github.com/user-attachments/assets/9165c00a-324e-473d-9b43-1c7e97f576a2)
you can find the application here...
https://github.com/Psynosaur/LVESS_485_Monitor/releases/tag/release-v1.07

lookout for the messages it sends in the debug console, you'll find

```bash
[2023-07-12 17:50:20.909] Send: AA 55 01 04 00 03 70 0D 0A
[2023-07-12 17:50:20.982] Recv: AA 55 01 04 48 0D 34 0F 0D 31 01 00 03 0D 31 0D 31 0D 31 0D 31 0D 31 0D 31 0D 31 0D 31 0D 33 0D 33 0D 32 0D 32 0D 33 0D 33 0D 34 0D 33 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 60 3E 0D 0A
```

I didn't ever use any registers like in MODBUS, I sent the payload in the via C# console app and got a response and parsed it.

**Comment by Psynosaur** [comment]
please use the new issue going forward, this original issue is related to the CAN bus not RS485.
